aidial_assistant/json_stream/json_object.py

```diff
@@ -55,47 +44,55 @@ async def get(self, key: str) -> JsonNode:
-    ):
-        normalised_stream = JsonNormalizer(stream)
-        char = await anext(normalised_stream)
-        if not char == JsonObject.token():
-            raise unexpected_symbol_error(char, stream.char_position)
-
-        separate = False
-        while True:
-            char = await normalised_stream.apeek()
-
-            if char == "}":
-                await normalised_stream.askip()
-                break
-
-            if char == ",":
-                if not separate:
-                    raise unexpected_symbol_error(char, stream.char_position)
-
-                await normalised_stream.askip()
-                separate = False
-            elif char == '"':
-                if separate:
-                    raise unexpected_symbol_error(char, stream.char_position)
-
-                key = await join_string(JsonString.read(stream))
-                colon = await anext(normalised_stream)
-                if not colon == ":":
-                    raise unexpected_symbol_error(colon, stream.char_position)
-
-                value = await dependency_resolver.resolve(stream)
-                await self.listener.put((key, value))
-                if isinstance(value, ComplexNode):
-                    await value.parse(stream, dependency_resolver)
-                separate = True
-            else:
+    @staticmethod
+    async def read(
+        stream: CharacterStream, dependency_resolver: NodeResolver
+    ) -> AsyncIterator[Tuple[str, JsonNode]]:
+        try:
+            normalised_stream = JsonNormalizer(stream)
+            char = await anext(normalised_stream)
+            if not char == JsonObject.token():
                 raise unexpected_symbol_error(char, stream.char_position)
 
-        await self.listener.put(None)
+            separate = False
+            while True:
+                char = await normalised_stream.apeek()
+
+                if char == "}":
+                    await normalised_stream.askip()
+                    break
+
+                if char == ",":
+                    if not separate:
+                        raise unexpected_symbol_error(
+                            char, stream.char_position
+                        )
+
+                    await normalised_stream.askip()
+                    separate = False
+                elif char == '"':
+                    if separate:
+                        raise unexpected_symbol_error(
+                            char, stream.char_position
+                        )
+
+                    key = await join_string(JsonString.read(stream))
+                    colon = await anext(normalised_stream)
+                    if not colon == ":":
+                        raise unexpected_symbol_error(
+                            colon, stream.char_position
+                        )
+
+                    value = await dependency_resolver.resolve(stream)
+                    yield key, value
+
+                    if isinstance(value, ReadableNode):
```

Comment:
You could move `read_to_end` for `JsonObject/Array/String` to `RootNodeResolver.resolve` to place them in the same method as `PrimitiveNode.collect(stream)` calls, which do the conceptually similar thing - move the stream pointer behind the entity which is being parsed.